examples/mcts.dx

```diff
@@ -0,0 +1,323 @@
+                child = from_just arr.(children.i)
+                (win, visits) = node_win_visit_stats child
+                uct win visits p_visits
+              ref := children.best_child
+              Continue
+
+' Expand the tree (in-place) by adding states one-move away from the current state.
+
+def mcts_expand {move state node_ix h}
+    (game: Game state move)
+    (tree_ref: Ref h (Tree node_ix (NodeStats state)))
+    (node: node_ix) : {State h} Unit =
+    -- assume `node` is a leaf node (base) and has a non-terminal state (base_state)
+
+    -- extract info about the base node
+    base_node = from_just $ get (fst_ref tree_ref)!node
+    (Branch base_stats base_parent _) = base_node
+    (_, _, base_level, base_state) = base_stats
+
+    -- prepare child nodes
+    (AsList n base_moves) = (get_at #possible_moves game) base_state
+    child_nodes = for i.
+      new_move = base_moves.i
+      new_state = (get_at #update game) base_state new_move
+      new_stats = (0, 0, base_level + 1, new_state)
+      Branch new_stats (Just node) (to_list [])
+
+    -- write to the tree
+    new_node = get (snd_ref tree_ref)
+    base_children_ix = for i:(Fin n). new_node +| (1 + ordinal i)
+    -- update base's children list
+    new_base_node = Branch base_stats base_parent (to_list base_children_ix)
+    (fst_ref tree_ref)!node := Just new_base_node
+    -- insert every child
+    for i.
+      child_ix = base_children_ix.i
+      (fst_ref tree_ref)!child_ix := Just child_nodes.i
+    snd_ref tree_ref := new_node +| n
+
+' Simulate the game starting at a given state until the game terminates and
+return the simulation result.
+
+def mcts_rollout {move state}
+    (game: Game state move)
+    (key: Key) (curr_state: state) : Bool =
+    -- output is w.r.t. current player
+    with_state curr_state \ref. iter \i.
+      curr_state' = get ref
+      if (get_at #game_over game) curr_state'
+        then
+          -- #win_or_lose is w.r.t. the last player
+          result = (get_at #win_or_lose game) curr_state'
+          Done $ xor (is_even i) result
+        else
+          (AsList _ new_moves) = (get_at #possible_moves game) curr_state'
+          -- rollout policy: uniform random
+          new_move = new_moves.(rand_idx (hash key i))
+          ref := (get_at #update game) curr_state' new_move
+          Continue
+
+' Update the record of a node and its chain of ancestors,
+based on the simulation result.
+
+def mcts_backpropagate {move state node_ix h} [Eq node_ix] 
+    (game: Game state move)
+    (tree_ref: Ref h (Tree node_ix (NodeStats state)))
+    (node: node_ix) (newResult: Float) : {State h} Unit =
+    -- newResult: win or lose w.r.t curr_player @node
+    leaf_min_player = get (fst_ref tree_ref)!node |>
+      from_just |> node_level |> is_odd
+    maxResult = if leaf_min_player 
+                    then (1. - newResult)
+                    else newResult
+
+    -- collect itself and all ancestors
```

Comment:
Question: Why collect the ancestors first and then loop over them to update?  Why not just update each ancestor as it's traversed, and dispense with the list entirely?

(Context: at the moment, `append` on a length-n list is O(n), not O(1), so collecting the list will be quadratic in the depth of the MCTS tree.)

Reply:
Done. 

You are definitely correct here. I was translating what I thought but having a list that grows in size is not a good idea.